File: nodejs/executor.js
// Paper-trading executor.
// Tracks open position, average entry, and realized PnL.
// Real execution: replace exchange.submitOrder() with a real venue call.
export class Executor {
  constructor(exchange, risk) {
    this.exchange = exchange;
    this.risk = risk;
    this.position = 0;
    this.avgEntry = 0;
    this.pnl = 0;
  }

  async submit(order) {
    if (!this.risk.check(this.position, order)) return null;

    const fill = await this.exchange.submitOrder(order);
    const signedQty = order.side === 'BUY' ? order.qty : -order.qty;

    // If this order reduces or flips the current position, book realized PnL.
    if (this.position !== 0 && Math.sign(signedQty) !== Math.sign(this.position)) {
      const closeQty = Math.min(Math.abs(signedQty), Math.abs(this.position));
      const direction = this.position > 0 ? 1 : -1;
      this.pnl += (fill.fillPrice - this.avgEntry) * closeQty * direction;
    }

    // Update average entry when adding to position; reset when flipping/opening.
    const newPos = this.position + signedQty;
    if (newPos !== 0 && (this.position === 0 || Math.sign(newPos) === Math.sign(this.position))) {
      this.avgEntry =
        (this.avgEntry * Math.abs(this.position) + fill.fillPrice * Math.abs(signedQty)) /
        Math.abs(newPos);
    } else {
      this.avgEntry = fill.fillPrice;
    }

    this.position = newPos;
    return fill;
  }
}

File: golang/executor.go
package main

import "math"

// Executor — paper trading.
// Tracks open position, average entry price, and realized PnL.
type Executor struct {
	exchange Exchange
	risk     *RiskManager
	Position float64
	AvgEntry float64
	Pnl      float64
}

func NewExecutor(ex Exchange, risk *RiskManager) *Executor {
	return &Executor{exchange: ex, risk: risk}
}

func (e *Executor) Submit(symbol string, sig *Signal) *Fill {
	order := Order{Symbol: symbol, Side: sig.Side, Qty: sig.Qty}
	if !e.risk.Check(e.Position, order) {
		return nil
	}

	fill := e.exchange.SubmitOrder(order)
	signedQty := order.Qty
	if order.Side == "SELL" {
		signedQty = -order.Qty
	}

	// Realize PnL when this order reduces/flips an open position.
	if e.Position != 0 && ((signedQty > 0) != (e.Position > 0)) {
		closeQty := math.Min(math.Abs(signedQty), math.Abs(e.Position))
		direction := 1.0
		if e.Position < 0 {
			direction = -1.0
		}
		e.Pnl += (fill.FillPrice - e.AvgEntry) * closeQty * direction
	}

	// Update avg entry when adding to position; reset on flip/open.
	newPos := e.Position + signedQty
	if newPos != 0 && (e.Position == 0 || ((newPos > 0) == (e.Position > 0))) {
		e.AvgEntry = (e.AvgEntry*math.Abs(e.Position) + fill.FillPrice*math.Abs(signedQty)) / math.Abs(newPos)
	} else {
		e.AvgEntry = fill.FillPrice
	}
	e.Position = newPos
	return &fill
}
